### PR TITLE
move _trial_temp to tox envtmpdir

### PR DESCRIPTION
**Pull request by graingert**
## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10036
* [ ] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [ ] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

More info here: https://freenode.logbot.info/twisted-dev/20201024#c5576741-c5577518

This helps enable a whole bunch of stuff:

* Running with `tox --parallel`
* avoiding problems like, pip failing with:

```
WARNING: Ignoring special file error '`/home/graingert/projects/twisted/_trial_temp/twisted.test.test_unix/DatagramUnixSocketTests/test_reprWithNewStyleProtocol/lk6x7h_1/temp` is a socket' encountered copying /home/graingert/projects/twisted/_trial_temp/twisted.test.test_unix/DatagramUnixSocketTests/test_reprWithNewStyleProtocol/lk6x7h_1/temp to /tmp/pip-req-build-1wyjs487/_trial_temp/twisted.test.test_unix/DatagramUnixSocketTests/test_reprWithNewStyleProtocol/lk6x7h_1/temp.
```

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -121,12 +121,12 @@ commands =
...
 
     ; Run tests without wrapping them using coverage.
-    nocov: python -m twisted.trial --reactor={env:TWISTED_REACTOR:default} --reporter={env:TRIAL_REPORTER:verbose}  {posargs:twisted}
+    nocov: python -m twisted.trial --temp-directory={envtmpdir}/_trial_temp --reactor={env:TWISTED_REACTOR:default} --reporter={env:TRIAL_REPORTER:verbose}  {posargs:twisted}
 
     ; Run the tests wrapped using coverage.
     withcov: python {toxinidir}/admin/_copy.py {toxinidir}/admin/zz_coverage.pth {envsitepackagesdir}/zz_coverage.pth
     withcov: coverage erase
-    withcov: coverage run -p --rcfile={toxinidir}/.coveragerc -m twisted.trial --reactor={env:TWISTED_REACTOR:default} --reporter={env:TRIAL_REPORTER:verbose} {posargs:twisted}
+    withcov: coverage run -p --rcfile={toxinidir}/.coveragerc -m twisted.trial --temp-directory={envtmpdir}/_trial_temp --reactor={env:TWISTED_REACTOR:default} --reporter={env:TRIAL_REPORTER:verbose} {posargs:twisted}
```
Comment (graingert) [inline tox.ini:129]: I think there's a way to avoid this duplication with some magic recursive configparser expansion

Comment (adiroiban) [reply tox.ini:129]: I think that is ok... we should not change this that often. thanks

**Review by adiroiban — APPROVED** [review]
Looks good. Thanks.

I guess that in these day everyone has a CPU with at least 2 cores.
Also the Azure / GiHub VMs have 2 cores.

When running trial, can we also enable 2 parallel runs `trial -j2 `?

For example on my laptop.

```
$ trial twisted.python.test.test_release  - takes 16 seconds
$ trial -j2 twisted.python.test.test_release - takes 5 seconds
```

So maybe we can enable j2 everywhere.

But this can be merged as it is and add j2 in a separate PR.

**Comment by adiroiban** [comment]
Another question... do we have the lists of tests that left temp files behind? I think that we should have separate PR to fix it.

Thanks

**Comment by graingert** [comment]
> Another question... do we have the lists of tests that left temp files behind? I think that we should have separate PR to fix it.
> 
> Thanks

All tests leave files behind, see 10037

**Pull request merged**